flow/layers/layer_state_stack.cc

```diff
@@ -0,0 +1,792 @@
+  if (builder) {
+    restore_count_ = builder->getSaveCount();
+    builder_ = builder;
+    reapply_all();
+  }
+}
+
+void LayerStateStack::set_delegate(MutatorsStack* stack) {
+  if (stack == mutators_) {
+    return;
+  }
+  clear_delegate();
+  if (stack) {
+    restore_count_ = stack->stack_count();
+    mutators_ = stack;
+    reapply_all();
+  }
+}
+
+void LayerStateStack::set_initial_cull_rect(const SkRect& cull_rect) {
+  FML_CHECK(is_empty()) << "set_initial_cull_rect() must be called before any "
+                           "state is pushed onto the state stack";
+  initial_cull_rect_ = cull_rect_ = cull_rect;
+}
+
+void LayerStateStack::set_initial_transform(const SkMatrix& matrix) {
+  FML_CHECK(is_empty()) << "set_initial_transform() must be called before any "
+                           "state is pushed onto the state stack";
+  initial_matrix_ = matrix_ = SkM44(matrix);
+}
+
+void LayerStateStack::set_initial_transform(const SkM44& matrix) {
+  FML_CHECK(is_empty()) << "set_initial_transform() must be called before any "
+                           "state is pushed onto the state stack";
+  initial_matrix_ = matrix_ = matrix;
+}
+
+void LayerStateStack::set_initial_state(const SkRect& cull_rect,
+                                        const SkMatrix& matrix) {
+  FML_CHECK(is_empty()) << "set_initial_state() must be called before any "
+                           "state is pushed onto the state stack";
+  initial_cull_rect_ = cull_rect_ = cull_rect;
+  initial_matrix_ = matrix_ = SkM44(matrix);
+}
+
+void LayerStateStack::set_initial_state(const SkRect& cull_rect,
+                                        const SkM44& matrix) {
+  FML_CHECK(is_empty()) << "set_initial_state() must be called before any "
+                           "state is pushed onto the state stack";
+  initial_cull_rect_ = cull_rect_ = cull_rect;
+  initial_matrix_ = matrix_ = matrix;
+}
+
+void LayerStateStack::reapply_all() {
+  // We use a local RenderingAttributes instance so that it can track the
+  // necessary state changes independently as they occur in the stack.
+  // Reusing |outstanding_| would wreak havoc on the current state of
+  // the stack. When we are finished, though, the local attributes
+  // contents should match the current outstanding_ values;
+  RenderingAttributes attributes = outstanding_;
+  SkM44 matrix = matrix_;
+  SkRect cull_rect = cull_rect_;
+  outstanding_ = {};
+  matrix_ = initial_matrix_;
+  cull_rect_ = initial_cull_rect_;
+  for (auto& state : state_stack_) {
+    state->reapply(this);
+  }
+  FML_DCHECK(attributes == outstanding_);
+  FML_DCHECK(matrix == matrix_);
+  FML_DCHECK(cull_rect == cull_rect_);
+}
+
+AutoRestore::AutoRestore(LayerStateStack* stack)
+    : layer_state_stack_(stack), stack_restore_count_(stack->stack_count()) {}
+
+AutoRestore::~AutoRestore() {
+  layer_state_stack_->restore_to_count(stack_restore_count_);
+}
+
+AutoRestore LayerStateStack::applyState(const SkRect& bounds,
```

Comment:
Inside the mutator we have "saveLayer" with the capitalization from everywhere else that method appears. Then we have its variant methods like "applyOpacity" which are meant to mirror it. This capitalization scheme then starts leaking into this code. I'm trying to figure out where the line is.

About all I've done so far is `get_draw_checkerboard` -> `draw_checkerboard` since it is an accessor, but unfortunately that makes it sound like an operation rather than a getter... Sigh. Maybe I'll rename it to "checkerboard_renderer".

Reply:
applyState could probably be renamed to resolve or protect as it conditionally applies a protective saveLayer only if there are outstanding incompatible attributes...?

Reply:
In `Impeller` it's `Canvas::SaveLayer`. We only have `saveLayer` from either Skia or legacy Blink code.